compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -128,6 +132,8 @@
     private Names names;
     private int finiteTypeCount = 0;
     private BUnionType expandedXMLBuiltinSubtypes;
+    private int recordCount = 0;
+    private SymbolEnv env;
```

Comment:
Types is like a stateless util class, so IMO, we shouldn't do this. Can't we pass the `env` from wherever the methods are called?

Reply:
Even with this approach, just setting `env` for match won't work well with other places that call `getTypeIntersection`.

For example, the following results in a bad, sad error due to `env` being null.

```ballerina
public function main() {
    map<int> m = {a: 1};

    if m is record {int a;} {
        
    }
}
```

```log
[2020-11-11 20:24:19,518] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.getTypeForRecordTypeMembersAssignableToType(Types.java:2813)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.getTypeForMapTypeMembersAssignableToType(Types.java:2804)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.lambda$getTypeIntersection$51(Types.java:3198)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.Types.getTypeIntersection(Types.java:3211)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeNarrower.visit(TypeNarrower.java:194)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangTypeTestExpr.accept(BLangTypeTestExpr.java:65)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeNarrower.analyzeExpr(TypeNarrower.java:224)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeNarrower.getNarrowedTypes(TypeNarrower.java:202)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeNarrower.evaluateTruth(TypeNarrower.java:92)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeNarrower.evaluateTruth(TypeNarrower.java:114)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:2282)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangIf.accept(BLangIf.java:56)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3047)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3015)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3011)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:392)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3047)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:364)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3047)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3015)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3007)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:280)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:254)
	at org.wso2.ballerinalang.compiler.CompilerDriver.typeCheck(CompilerDriver.java:347)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:276)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:268)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:144)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:176)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:221)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:101)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:109)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:55)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:297)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at org.ballerinalang.tool.Main.main(Main.java:57)
```

Reply:
I think we'll have to fix this by making `getTypeIntersection` accept the `env`.

Reply:
Changed